projects/knora-ui/src/lib/viewer/values/link-value/link-value.component.ts

```diff
@@ -0,0 +1,169 @@
+import {Component, Inject, Input, OnChanges, OnDestroy, OnInit, SimpleChanges} from '@angular/core';
+import {BaseValueComponent} from '../base-value.component';
+import {CreateLinkValue, ReadLinkValue, ReadResource, UpdateLinkValue, KnoraApiConnection} from '@knora/api';
+import {Subscription} from 'rxjs';
+import {AbstractControl, FormBuilder, FormControl, FormGroup, ValidatorFn} from '@angular/forms';
+import {KnoraApiConnectionToken} from '../../../core';
+@Component({
+  selector: 'kui-link-value',
+  templateUrl: './link-value.component.html',
+  styleUrls: ['./link-value.component.scss']
+})
+export class LinkValueComponent extends BaseValueComponent implements OnInit, OnChanges, OnDestroy {
+  @Input() displayValue?: ReadLinkValue;
+  @Input() parentResource: ReadResource;
+  @Input() propType: string;
+  resources: ReadResource[];
+  restrictToResourceClass: string;
+  valueFormControl: FormControl;
+  commentFormControl: FormControl;
+  form: FormGroup;
+
+  valueChangesSubscription: Subscription;
+  // label cannot contain logical operations of lucene index
+  customValidators = [];
+
+  constructor(@Inject(FormBuilder) private fb: FormBuilder, @Inject(KnoraApiConnectionToken) private knoraApiConnection: KnoraApiConnection) {
+    super();
+  }
+
+  /**
+   * Displays a selected resource using its label.
+   *
+   * @param resource the resource to be displayed (or no selection yet).
+   * @returns
+   */
+  displayResource(resource: ReadResource | null): string {
+
+    // null is the initial value (no selection yet)
+    if (resource !== null) {
+      return resource.label;
+    }
+  }
+
+  /**
+   * Search for resources whose labels contain the given search term, restricting to to the given properties object constraint.
+   * this is to be used for update and new linked resources
+   *
+   * @param label to be searched
+   */
+  searchByLabel(searchTerm: string): ReadResource[] {
+    const linkType = this.parentResource.getLinkPropertyIriFromLinkValuePropertyIri(this.propType);
+    if (typeof linkType === 'string') {
+      this.restrictToResourceClass = this.parentResource.entityInfo.properties[linkType].objectType;
```

Comment:
I think object type is optional. If not defined, you can omit the restriction.

Reply:
I thought objectType should be definitely specified for link properties.

Reply:
According to [documentation](https://docs.knora.org/paradox/02-knora-ontologies/knora-base.html#constraints-on-the-types-of-property-subjects-and-objects), subjectClassConstraing is optional for linkValues but ObjectClassConstraint is mandatory. 
@benjamingeer I remember this from the time I was writing ontologies. Is my understanding still correct?

Reply:
> According to [documentation](https://docs.knora.org/paradox/02-knora-ontologies/knora-base.html#constraints-on-the-types-of-property-subjects-and-objects), subjectClassConstraing is optional for linkValues but ObjectClassConstraint is mandatory.

I think `objectType` optional in `ResourcePropertyDefinition` because there are properties that do not have an object class constraint (`ResourcePropertyDefinition` has to represent any property). Because of that, it's optional for all properties in `ResourcePropertyDefinition`.

Reply:
Which resource properties have no object class constraint? Maybe I’m forgetting something, but I think there shouldn’t be any.

Reply:
I think it is the `resourceProperty`:

```
###  http://www.knora.org/ontology/knora-base#resourceProperty

:resourceProperty rdf:type owl:ObjectProperty ;

          rdfs:comment "The base property of properties that point from Knora resources to Knora resources or values. These properties are required to have cardinalities in the resource classes in which they are used."@en ;

          :subjectClassConstraint :Resource .
```

Reply:
and then in the v2 ontology:

```json
{
    "@id" : "knora-api:resourceProperty",
    "@type" : "owl:ObjectProperty",
    "knora-api:isResourceProperty" : true,
    "knora-api:subjectType" : {
      "@id" : "knora-api:Resource"
    },
    "rdfs:comment" : "The base property of properties that point from Knora resources to Knora resources or values. These properties are required to have cardinalities in the resource classes in which they are used."
  }
```

Reply:
But `knora-api:resourceProperty` can never be used in a resource. It's only an abstract base property. Would it help if we changed it to have `"knora-api:isResourceProperty" : false`?

Reply:
The thing is that I have to parse it and turn into a class. And if that class had a non optional objectType, it would fail.

Reply:
You said:

> `objectType` optional in `ResourcePropertyDefinition` because there are properties that do not have an object class constraint

I think this doesn't make sense. All resource properties have an `objectType`. If the only exception is `knora-api:resourceProperty`, which is not really a resource property but only the base property of all resource properties, we can change `knora-api:resourceProperty` so it's not considered a resource property. Then you can make `objectType` required instead of optional in `ResourcePropertyDefinition`.

Reply:
So then we would not turn `knora-api:isResourceProperty` into a `ResourcePropertyDefinition`?

Reply:
> So then we would not turn knora-api:isResourceProperty into a ResourcePropertyDefinition?

Do you mean `knora-api:resourceProperty` or `knora-api:isResourceProperty`?

Reply:
> > So then we would not turn knora-api:isResourceProperty into a ResourcePropertyDefinition?
> 
> Do you mean `knora-api:resourceProperty` or `knora-api:isResourceProperty`?

I mean `knora-api:resourceProperty`.

If you set `knora-api:isResourceProperty` to `false` for `knora-api:resourceProperty`, then it won't be converted into a `ResourcePropertyDefinition` (I think then it will be just ignored).

https://github.com/dasch-swiss/knora-api-js-lib/blob/a99e32d0c6f7b8f12e2a952716c5270d8999e9f0/src/models/v2/ontologies/OntologyConversionUtil.ts#L73-L81

I suggest to create issues for this in knora-api and knora-api-js-lib and do it when we have time for it.

Reply:
Or should it be handled like a system property?

Reply:
> If you set knora-api:isResourceProperty to false for knora-api:resourceProperty, then it won't be converted into a ResourcePropertyDefinition (I think then it will be just ignored).

Yes, I think that would make sense. The client code should always be able to trust that a resource property will have an object type.

> I suggest to create issues for this in knora-api and knora-api-js-lib and do it when we have time for it.

OK, will do.

Reply:
system properties are expected to to have the `knora-api:isResourceProperty` flag:

https://github.com/dasch-swiss/knora-api-js-lib/blob/a99e32d0c6f7b8f12e2a952716c5270d8999e9f0/src/models/v2/ontologies/OntologyConversionUtil.ts#L83-L91

Reply:
> Or should it be handled like a system property?

It's never used in resources. It's only there as documentation. So I think client code will never use it.

Reply:
If I remember correctly, `knora-api:isResourceProperty` is absent or present. If present, it is set to `true`. Or am I mistaken?

Reply:
absence is like implicitly `false`

Reply:
> system properties are expected to not have the knora-api:isResourceProperty flag:

That makes sense. System properties are not resource properties. The idea of a resource property is that its object is a `knora-api:Value` or a `knora-api:Resource`.

> If I remember correctly, knora-api:isResourceProperty is absent or present. If present, it is set to true.

Yes.

> absence is like implicitly false

Yes.